test/behavior/cast.zig

```diff
@@ -1622,3 +1622,201 @@ test "coercion from single-item pointer to @as to slice" {
 
     try expect(t[0] == 1);
 }
+
+test "peer type resolution: const sentinel slice and mutable non-sentinel slice" {
+    if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_arm) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_wasm) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_sparc64) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_spirv64) return error.SkipZigTest; // TODO
+
+    const S = struct {
+        fn doTheTest(comptime T: type, comptime s: T) !void {
+            var a: [:s]const T = undefined;
+            var b: []T = undefined;
+            try expect(@TypeOf(a, b) == []const T);
+            try expect(@TypeOf(b, a) == []const T);
```

Comment:
```suggestion
            comptime assert(@TypeOf(a, b) == []const T);
            comptime assert(@TypeOf(b, a) == []const T);
```

In all of these new behavior tests, please actually test that the values behave correctly as well. In #15709, I gave you a nice test case that did this already.